### DISCUSSION TITLE
Any Plan to Provide React Native Support?

### DISCUSSION THREAD

**Discussion by ThaminduDilshan** [panva/jose#332] [Q&A]
I'm using some auth node module in my react native application which is using `jose` as a dependency. I get the following error which I believe is due to the fact that in-build typescript/ node libraries are not supported by React native.

```
ERROR    ReferenceError: Can't find variable: crypto
ERROR    Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication)
```

Even using [node-libs-react-native](https://github.com/parshap/node-libs-react-native) in my app didn't fix the issue. It appears adding a resolver to `metro.config.js` to provide extra node modules doesn't fix the issue for the libraries used by dependencies.

Do you have any plan to provide native support for jose or are there any workarounds to get this fixed?

**Comment by panva** [comment]
> Any Plan to Provide React Native Support?

Countless Web APIs missing in the react native runtime, most importantly, Web Cryptography API. So, no plan.

**Comment by Sytten** [comment]
Has there been any progress on the react-native side over the last year?

**Reply by panva** [reply]
Did react-native ship any Web APIs? Specifically Web Cryptography API?

**Reply by Sytten** [reply]
No they said a while ago that they dont want to increase RN surface API so it would need to be a community library.
One exists based on crypto.js but it doesn't implement all the APIs (especially not the newer ones related to eddsa) and it's not native (so likely the CPRNG isn't very good) so I guess we are pretty stuck.
I am quite stunned that this is still an issue that many years after the release of RN.
What would be possible in the meantime would be to use a library like libsodium or Themis that do have RN bindings but not the same API as the crypto API (https://github.com/lyubo/react-native-sodium). The other is to use an hidden webview to access the crypto web API (https://www.npmjs.com/package/react-native-webview-crypto).

**Reply by panva** [reply]
> The other is to use an hidden webview to access the crypto web API (https://www.npmjs.com/package/react-native-webview-crypto).

So you can use the same pattern to access the browser distro of jose?

**Reply by Sytten** [reply]
I guess you can probably do that yeah I have not tried it, just read about it. It seems pretty hacky but I guess it would work.

**Reply by panva** [reply]
Still better crypto than crappy polyfills.

**Reply by Sytten** [reply]
Agreed, I also saw https://github.com/ghbutton/react-native-simple-crypto which is a native implementation of a couple of APIs. That would be an alternative.
